### PR TITLE
Add leave room warning for last admin

### PR DESCRIPTION
**Pull request by Arnei**
If the last room administrator leaves a room, other users cannot gain admin privilges anymore, leaving the room in an unmoderable state. To help in avoiding this scenario without actually preventing an admin from leaving the room if they really want, this commit adds a new warning message.

Attempts to help with: https://github.com/vector-im/element-web/issues/2855

Screenshots as recommended by robintown

With no warnings:
Screenshot 2024-03-21 at 11 53 35

With warnings:
Screenshot 2024-03-21 at 11 53 49

Signed-off-by: Arne Wilken arnepokemon@yahoo.de

type: enhancement

---
Here's what your changelog entry will look like:

## ✨ Features
 * Add leave room warning for last admin ([\#9452](https://github.com/matrix-org/matrix-react-sdk/pull/9452)). Contributed by @Arnei.

### CODE REVIEW

**Comment by t3chguy** [comment]
Hey @Arnei thanks for your contribution, any chance of satisfying the code coverage test requirement?

**Comment by t3chguy** [comment]
Actually I take it back, MatrixChat is currently too indebted to be tested, you can skip it for now. The web app team have plans to fix the situation. Just please resolve merge conflicts.

File: src/components/structures/MatrixChat.tsx [hunk 1]
```diff
@@ -1132,6 +1133,29 @@ export default class MatrixChat extends React.PureComponent<IProps, IState> {
                 ));
             }
         }
+
+        const client = MatrixClientPeg.get();
+        const plEvent = roomToLeave.currentState.getStateEvents(EventType.RoomPowerLevels, '');
+        const plContent = plEvent ? (plEvent.getContent() || {}) : {};
```
Comment (robintown) [inline src/components/structures/MatrixChat.tsx:1139]:
```suggestion
        const plContent = plEvent ? plEvent.getContent() : {};
```

`getContent` already does the `|| {}` step

File: src/components/structures/MatrixChat.tsx [hunk 2]
```diff
@@ -1132,6 +1133,29 @@ export default class MatrixChat extends React.PureComponent<IProps, IState> {
...
+            const maxUserLevel = Math.max(...userLevelValues);
+            // If the user is the only user with highest power level
+            if (maxUserLevel === currentUserLevel &&
+                userLevelValues.lastIndexOf(maxUserLevel) == userLevelValues.indexOf(maxUserLevel)) {
+                warnings.push((
+                    <span className="warning" key="last_admin_warning">
+                        { ' '/* Whitespace, otherwise the sentences get smashed together */ }
+                        { _t("You are the sole person with the highest role in this room. " +
+                        "If you leave, the room could become unmoderable. Consider giving " +
+                        "another user your role.") }
```
Comment (robintown) [inline src/components/structures/MatrixChat.tsx:1153]: I think we should hide this warning if the "You are the only person here" warning is shown

**Review by robintown — COMMENTED** [review]
This will need a signoff from the design team - could you please add a screenshot of the warning to make it easier for them to review the copy?

File: src/components/structures/MatrixChat.tsx (hunk 2 above)
Comment (Arnei) [reply src/components/structures/MatrixChat.tsx:1153]: That is already the case. `leaveRoomWarnings(roomId: string)` returns early if the "You are the only person here" warning is added, so no further warnings are shown in that case.

**Review by robintown — APPROVED** [review]
The code looks good, just waiting on design now.

**Comment by Arnei** [comment]
Would it be rude to ping Design for their opinion?

File: src/components/structures/MatrixChat.tsx [hunk 3]
```diff
@@ -1132,6 +1133,29 @@ export default class MatrixChat extends React.PureComponent<IProps, IState> {
...
+        if (isNumberArray(userLevelValues)) {
+            const maxUserLevel = Math.max(...userLevelValues);
+            // If the user is the only user with highest power level
+            if (maxUserLevel === currentUserLevel &&
+                userLevelValues.lastIndexOf(maxUserLevel) == userLevelValues.indexOf(maxUserLevel)) {
+                warnings.push((
+                    <span className="warning" key="last_admin_warning">
+                        { ' '/* Whitespace, otherwise the sentences get smashed together */ }
+                        { _t("You are the sole person with the highest role in this room. " +
+                        "If you leave, the room could become unmoderable. Consider giving " +
```
Comment (MadLittleMods) [inline src/components/structures/MatrixChat.tsx:1152]: `unmoderable` doesn't feel like a word. `unmoderatable` sounds better to me but I also don't see it as a real word so probably need to rearrange the prose to accommodate proper English. Perhaps:

"If you leave, the room might not be able to be moderated anymore."

**Comment by daniellekirkwood** [comment]
Discussed the content with @aaronraimist and here's our suggestion... 
--- 
**Headline: Sure you want to leave?**

Body copy: You’re the only <administrator/moderator> in this room. If you leave, nobody will be able to change room settings or take other important actions.

Body copy: Consider making someone else <an administrator/ a moderator> before you go.
--- 
We'd use this copy for both cases as an admin or moderator leaving a private room should know that it's private and should know the circumstances needed for re-entry.

It would also be great if we could make that button red instead of green :)

**Comment by aaronraimist** [comment]
@daniellekirkwood I think you probably meant to @ the Aaron that works at Element :)

**Comment by dbkr** [comment]
I've resolved the conflicts and incorporated the copy from the feedback, with separate text for the admin / mod case. I left the title as-is because I think this might need splitting out from the general "leave room" action, and not sure if it also wants to apply to the case when there aren't warnings?

The type utils file got deleted and we just don't have these kind of type checks in react-sdk (for better or worse) so I've just inlined it.

Also put it in strong tags for consistency with the other warnings. Screenshots updated. Probably wants eyes from someone else to check my work.

**Comment by richvdh** [comment]
@dbkr can you update the screenshots?

**Comment by dbkr** [comment]
Screenshots should be updated? And yeah, seems like we'd already decided it was okay without the coverage tests.

**Pull request merged**